File: jianzhioffer/firstround/jianzhi_51.py
"""
方法一：暴力解，时间复杂度O(n^2)，空间复杂度O(1)，会超时
"""
# class Solution(object):
#     def reversePairs(self, nums):
#         """
#         :type nums: List[int]
#         :rtype: int
#         """
#         if not nums:
#             return 0
#         res = 0
#         for i in range(len(nums)-1):
#             for j in range(i+1, len(nums)):
#                 if nums[j] < nums[i]:
#                     res += 1
#         return res


"""
方法二：归并排序，时间复杂度O(nlogn)(同归并排序)，空间复杂度O(n)(同归并排序，建立了一个辅助数组tmp来存储排序后的元素)
"""
class Solution(object):
    def mergeSort(self, array, tmp, left, right):
        if left >= right:
            return 0
        mid = (left + right) // 2
        inv_count = self.mergeSort(array, tmp, left, mid) + self.mergeSort(array, tmp, mid+1, right)
        i, j, pos = left, mid+1, left
        while i <= mid and j <= right:
            if array[i] <= array[j]:  # 只在i右移时才计算array[i]的逆序对数，因为i右移表明array[i]<=array[j]，
                tmp[pos] = array[i]  # 但array[i]比array[mid+1,...,j-1]都要大，所以array[i]的逆序对数=j-(mid+1)
                i += 1
                inv_count += (j - (mid + 1))
            else:
                tmp[pos] = array[j]
                j += 1
            pos += 1
        for k in range(i, mid+1):  # 如果i<mid+1，说明前半部分剩下的元素比后半部分所有元素都要大，
            tmp[pos] = array[k]  # 所以前半部分剩下的元素的逆序对数都等于j-(mid+1)，此时该式的值等于后半部分的元素个数
            pos += 1
            inv_count += j - (mid + 1)
        for k in range(j, right+1):
            tmp[pos] = array[k]
            pos += 1
        array[left:right+1] = tmp[left:right+1]
        return inv_count

    def reversePairs(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        n = len(nums)
        tmp = [0] * n
        res = self.mergeSort(nums, tmp, 0, n-1)
        return res


if __name__ == '__main__':
    nums = [7,5,6,4]
    obj = Solution()
    result = obj.reversePairs(nums)
    print(result)

File: jianzhioffer/secondround/jianzhi_51.cpp
#include <vector>
using namespace std;

class Solution
{
public:
    Solution();
    ~Solution();
    int reversePairs(vector<int>& nums);

private:
    int mergeSort(vector<int>& nums, vector<int>& tmp, int left, int right);
};

Solution::Solution() {}

Solution::~Solution() {}


//归并排序。时间复杂度O(NlogN)，空间复杂度O(N)
int Solution::reversePairs(vector<int>& nums)
{
    if(nums.empty()) return 0;
    int n = nums.size();
    vector<int> tmp(n);
    return mergeSort(nums, tmp, 0, n-1);
}

int Solution::mergeSort(vector<int>&nums, vector<int>& tmp, int left, int right)
{
    if(left>=right) return 0;

    int mid = (left+right)/2;
    int reverse_count = mergeSort(nums, tmp, left, mid) + mergeSort(nums, tmp, mid+1, right);
    int i = left, j = mid+1;
    int pos = left;
    while (i<=mid && j<=right)
    {
        if(nums[i]<=nums[j])
        {
            tmp[pos++] = nums[i];
            i++;
            reverse_count += j - (mid+1);
        }
        else
        {
            tmp[pos++] = nums[j];
            j++;
        }
    }
    
    for (int k = i; k <= mid; k++)
    {
        tmp[pos++] = nums[k];
        reverse_count += j - (mid+1);
    }
    for (int k = j; k <= right; k++)
    {
        tmp[pos++] = nums[k];
    }
    copy(tmp.begin()+left, tmp.begin()+right+1, nums.begin()+left);
    return reverse_count;
}

// //暴力法。时间复杂度O(N^2)，空间复杂度O(1)
// int Solution::reversePairs(vector<int>& nums)
// {
//     if(nums.empty() || nums.size()==1) return 0;
//     int res = 0;
//     for (int i = 0; i < nums.size()-1; i++)
//     {
//         for (int j = i+1; j < nums.size(); j++)
//         {
//             if(nums[i]>nums[j]) res++;
//         }
//     }
//     return res;
// }

int main(int argc, char const *argv[])
{
    vector<int> nums={7,5,6,4};
    Solution obj;
    int result = obj.reversePairs(nums);
    return 0;
}
